sound/soc/amd/acp/acp-sdw-sof-mach.c

```diff
@@ -0,0 +1,1300 @@
+				return -EINVAL;
+			}
+
+			ret = fill_sdw_codec_dlc(dev, adr_link_next,
+						 &codecs[codec_dlc_index],
+						 j, dai_index);
+			if (ret)
+				return ret;
+
+			codec_conf[*codec_conf_index].dlc = codecs[codec_dlc_index];
+			codec_conf[*codec_conf_index].name_prefix =
+					adr_link_next->adr_d[j].name_prefix;
+			dev_err(dev, "codec_conf_index:%d name_prefix:%s\n", *codec_conf_index,
+				codec_conf[*codec_conf_index].name_prefix);
+
+			codec_dlc_index++;
+			(*codec_conf_index)++;
+		}
+		j = 0;
+
+		/* check next link to create codec dai in the processed group */
+		i++;
+	}
+
+	  /* find codec info to create BE DAI */
+	codec_index = find_codec_info_part(adr_link->adr_d[adr_index].adr);
+	if (codec_index < 0)
+		return codec_index;
+	codec_info = &codec_info_list[codec_index];
+
+	if (codec_info->ignore_acp_dmic)
+		*ignore_acp_dmic = true;
+
+	sdw_link_id = (adr_link->adr_d[adr_index].adr) >> 48;
+	for_each_pcm_streams(stream) {
+		struct snd_soc_dai_link_ch_map *sdw_codec_ch_maps;
+		char *name, *cpu_name;
+		int playback, capture;
+		static const char * const sdw_stream_name[] = {
+			"SDW%d-PIN%d-PLAYBACK",
+			"SDW%d-PIN%d-CAPTURE",
+			"SDW%d-PIN%d-PLAYBACK-%s",
+			"SDW%d-PIN%d-CAPTURE-%s",
+		};
+
+		if (!codec_info->dais[dai_index].direction[stream])
+			continue;
+
+		*be_id = codec_info->dais[dai_index].dailink[stream];
+		if (*be_id < 0) {
+			dev_err(dev, "Invalid dailink id %d\n", *be_id);
+			return -EINVAL;
+		}
+
+		sdw_codec_ch_maps = devm_kcalloc(dev, codec_num,
+						 sizeof(*sdw_codec_ch_maps), GFP_KERNEL);
+		if (!sdw_codec_ch_maps)
+			return -ENOMEM;
+
+		switch (sdw_link_id) {
+		case AMD_SDW0:
+			switch (*be_id) {
+			case SDW_JACK_OUT_DAI_ID:
+				cpu_pin_id = SW0_AUDIO0_TX;
+				break;
+			case SDW_JACK_IN_DAI_ID:
+				cpu_pin_id = SW0_AUDIO0_RX;
+				break;
+			case SDW_AMP_OUT_DAI_ID:
+				cpu_pin_id = SW0_AUDIO1_TX;
+				break;
+			case SDW_AMP_IN_DAI_ID:
+				cpu_pin_id = SW0_AUDIO1_RX;
+				break;
+			case SDW_DMIC_DAI_ID:
+				cpu_pin_id = SW0_AUDIO2_RX;
+				break;
+			default:
+				dev_err(dev, "Invalid be id:%d\n", *be_id);
+				return -EINVAL;
+			}
+			break;
+		case AMD_SDW1:
+			switch (*be_id) {
+			case SDW_JACK_OUT_DAI_ID:
+			case SDW_AMP_OUT_DAI_ID:
+				cpu_pin_id = SW1_AUDIO0_TX;
+				break;
+			case SDW_JACK_IN_DAI_ID:
+			case SDW_AMP_IN_DAI_ID:
+			case SDW_DMIC_DAI_ID:
+				cpu_pin_id = SW1_AUDIO0_RX;
+				break;
+			default:
+				dev_err(dev, "invalid be_id:%d\n", *be_id);
+				return -EINVAL;
+			}
+			break;
+		default:
+			dev_err(dev, "Invalid link id:%d\n", sdw_link_id);
+			break;
+		}
+
+		/* create stream name according to first link id */
+		if (append_dai_type) {
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      sdw_stream_name[stream + 2], sdw_link_id, cpu_pin_id,
+					      type_strings[codec_info->dais[dai_index].dai_type]);
+		} else {
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      sdw_stream_name[stream], sdw_link_id, cpu_pin_id);
+		}
```

Comment:
Since you guys presumably don't have legacy stuff to support here, would it not be simpler to just always append the dai type?

Reply:
Although I guess this way might be nicer if the plan is to try and share the code with Intel.

Reply:
Currently our Linux solution is not enabled in any of OEM platforms.
As per our understanding , appending_dai_type is required when multiple peripheral devices connected on the same link. We still need legacy stuff here.

Reply:
@charleskeepax : We need to provide support for legacy stack (No DSP enabled) as well. Once our SOF based machine driver changes finalized, we will work on implementing AMD SoundWire legacy machine driver.
SoundWire DAI creation logic will also change  based on ACP family version. Current machine driver is targeted for ACP v6.3 Variant platforms.